routers/api/packages/npm/npm.go

```diff
@@ -160,6 +161,43 @@ func UploadPackage(ctx *context.Context) {
 		}
 	}
 
+	if npmPackage.Metadata.Repository.Type == "git" {
+		repoURL := npmPackage.Metadata.Repository.URL
```

Comment:
`repository` can also appear in string form, e.g. `github:user/repo`, does this trip the parser? I think you could detect for example `my.domain/owner/repo` syntax as well from that string.

Reply:
The string form is only your local file; npm & yarn only send them in their object form, i.e. `{ type: xxx, url: yyy }`.

Reply:
I wouldn't rely on that behaviour, there may be other package managers that may follow the package.json spec more exactly and not do this transformation. We should follow the spec.

Reply:
Relying in that is following the spec; the registry API is actually not simply uploading the package.json; see https://github.com/npm/registry/blob/master/docs/REGISTRY-API.md. It also states that a packages 'repositoy' field when publishing is an object with `type` and `url`.

I think the string-only variant is just an convenience provided by frontends (like yarn & the npm cli).

Reply:
Ok, seems fine then.